homeassistant/components/simplisafe/__init__.py

```diff
@@ -0,0 +1,133 @@
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.alarm_control_panel import (
+    SCAN_INTERVAL as DEFAULT_SCAN_INTERVAL)
+from homeassistant.config_entries import SOURCE_IMPORT
+from homeassistant.const import (
+    CONF_CODE, CONF_PASSWORD, CONF_SCAN_INTERVAL, CONF_USERNAME)
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.util.json import load_json, save_json
+
+from homeassistant.helpers import config_validation as cv
+
+from .config_flow import configured_instances
+from .const import (
+    DATA_CLIENT, DATA_FILE_SCAFFOLD, DATA_LISTENER, DOMAIN, TOPIC_UPDATE)
+
+CONF_TOKEN_FILE = 'token_file'
+
+REQUIREMENTS = ['simplisafe-python==3.1.6']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ACCOUNTS = 'accounts'
+
+ACCOUNT_CONFIG_SCHEMA = vol.Schema({
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_CODE): cv.string,
+    vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+        cv.time_period
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_ACCOUNTS):
+            vol.All(cv.ensure_list, [ACCOUNT_CONFIG_SCHEMA]),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the SimpliSafe component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    for account in conf[CONF_ACCOUNTS]:
+        if account[CONF_USERNAME] in configured_instances(hass):
+            continue
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: account[CONF_USERNAME],
+                    CONF_PASSWORD: account[CONF_PASSWORD],
+                    CONF_CODE: account.get(CONF_CODE),
+                }))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up OpenUV as config entry."""
+    from simplipy import API
+    from simplipy.errors import SimplipyError
+
+    token_filepath = hass.config.path(
+        DATA_FILE_SCAFFOLD.format(config_entry.data[CONF_USERNAME]))
+    token_data = await hass.async_add_executor_job(load_json, token_filepath)
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        simplisafe = await API.login_via_token(
+            token_data['refresh_token'], websession)
+    except SimplipyError as err:
+        _LOGGER.error("There was an error during setup: %s", err)
+        return False
+
+    await hass.async_add_executor_job(
+        save_json, token_filepath, {'refresh_token': simplisafe.refresh_token})
```

Comment:
We just saved this in the config flow, do we need to save it again here?

Reply:
Consider the case after the config entry has been saved. When this function is run again, `API.login_via_token` is called, which uses the existing refresh token and generates a new one – I'm saving that new one.

Reply:
we shouldn't store any JSON. The config entry is where we store auth related data. You can update the data in a config entry too. 

If data is not related to authentication, we use the storage helper. We should never call `save_json` directly anymore.